### PR TITLE
[SPARK-43799][PYTHON] Add descriptor binary option to Pyspark Protobuf API

### PR DESCRIPTION
**Pull request by rangadi**
### What changes were proposed in this pull request?
This updated Protobuf Pyspark API to allow passing binary FileDescriptorSet rather than a file name. This is a Python follow up to feature implemented in Scala in #41192. 

### Why are the changes needed?
  - This allows flexibility for Pyspark users to provide binary descriptor set directly.
  - Even if users are using file path, Pyspark avoids passing file name to Scala and reads the descriptor file in Python. This avoids having to read the file in Scala. 

### Does this PR introduce _any_ user-facing change?

  - This adds extra arg to `from_protobuf()` and `to_protobuf()` API. 

### How was this patch tested?
  - Doc tests
  - Manual tests

### CODE REVIEW

**Comment by rangadi** [comment]
cc: @LuciferYang, @pengzhon-db

**Review by HyukjinKwon — APPROVED** [review]
Looks good from a cursory look

File: python/pyspark/sql/protobuf/functions.py [hunk 1]
```diff
@@ -61,9 +65,11 @@ def from_protobuf(
         The Protobuf class name when descFilePath parameter is not set.
         E.g. `com.example.protos.ExampleEvent`.
     descFilePath : str, optional
-        The protobuf descriptor file.
+        The Protobuf descriptor file.
     options : dict, optional
         options to control how the protobuf record is parsed.
+    binaryDescriptorSet: bytes, optional
+        The Protobuf `FileDescriptorSet` serialized as binary.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/protobuf/functions.py:72]: Actually we should add `versionchanged:: 3.5.0` here.

File: python/pyspark/sql/protobuf/functions.py [hunk 2]
```diff
@@ -232,6 +274,12 @@ def to_protobuf(
     return Column(jc)
 
 
+def _read_descriptor_set_file(filePath: str) -> bytes:
+    # TODO: Throw structured errors like "PROTOBUF_DESCRIPTOR_FILE_NOT_FOUND" etc.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/protobuf/functions.py:278]: and should file TODO'ed JIRA such as `TODO(SPARK-XXXXX)`. It's a bit trouble some to file for every todo but we found that some of todos are just left out, forgotten, and nobody remembers what it was :-).

File: python/pyspark/sql/protobuf/functions.py (hunk 1 above)
Comment (rangadi) [reply python/pyspark/sql/protobuf/functions.py:75]: Good call. Updated.

File: python/pyspark/sql/protobuf/functions.py (hunk 2 above)
Comment (rangadi) [reply python/pyspark/sql/protobuf/functions.py]: Agree. Added jira ticket and updated the comment.
Can we throw structured error now in Python? In that case, I can actually fix it in this PR itself. 
Please point to an example.

**Comment by rangadi** [comment]
The tests should pass in an hour or so. Please merge it you are around.

**Comment by rangadi** [comment]
@LuciferYang Please merge this when you get a chance.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by rangadi** [comment]
Thanks @HyukjinKwon!

**Pull request closed without merging**